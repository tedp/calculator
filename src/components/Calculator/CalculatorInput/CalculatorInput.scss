@import "../../../styles/styles.scss";

$module: "CalculatorInput";

.#{$module} {
  .top-panel {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;

    .Button {
      border: 1px solid $mid-grey;

      background-color: $dark-grey;
      &:hover {
        background-color: $dark-grey;
        border-color: $light-grey;
      }
    }
  }

  .centre-panel {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;

    .Button {
      border: 1px solid $dark-grey;
      background-color: #f9f9f9;
      &:hover {
        background-color: #f1f1f1;
        border-color: $mid-grey;
      }
    }
  }

  .bottom-panel {
    display: flex;
    justify-content: space-around;

    .Button {
      flex-basis: 93%;
      background-color: $equals-background-color;
      border: none;

      &:hover {
        background-color: $equals-hover-color;
        @include box-shadow();
        border-color: $equals-hover-border-color;
      }

      &:active {
        font-weight: bold;
      }
    }
  }
}
