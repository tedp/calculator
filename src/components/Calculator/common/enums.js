const KeyType = Object.freeze({
  numeric: 'numeric',
  dot: 'dot',
  operator: 'operator',
  clear: 'clear',
  result: 'result',
});

const DisplayState = Object.freeze({
  numeric: 'numeric',
  decimal: 'decimal',
  dot: 'dot',
  operator: 'operator',
});

const CalcState = Object.freeze({
  resulted: 'resulted',
  cleared: 'cleared',
  input: 'input',
});

export { KeyType, CalcState, DisplayState };
