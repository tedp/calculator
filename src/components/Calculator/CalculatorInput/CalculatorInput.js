import React from 'react';
import './CalculatorInput.scss';
import { KeyType } from '../common/enums';
import Button from '../Button/Button';

class CalculatorInput extends React.Component {
  createButtons(buttons) {
    return buttons.map((button) => (
      <Button
        key={button.keyValue}
        value={button.keyValue}
        onClick={() => this.props.onInput(button)}
      />
    ));
  }

  render() {
    const topPanelButtons = [
      { keyValue: '+', keyType: KeyType.operator },
      { keyValue: '-', keyType: KeyType.operator },
      { keyValue: '\u00d7', keyType: KeyType.operator },
      { keyValue: '\u00f7', keyType: KeyType.operator },
    ];
    const centrePanelButtons = [
      { keyValue: '9', keyType: KeyType.numeric },
      { keyValue: '8', keyType: KeyType.numeric },
      { keyValue: '7', keyType: KeyType.numeric },
      { keyValue: '6', keyType: KeyType.numeric },
      { keyValue: '5', keyType: KeyType.numeric },
      { keyValue: '4', keyType: KeyType.numeric },
      { keyValue: '3', keyType: KeyType.numeric },
      { keyValue: '2', keyType: KeyType.numeric },
      { keyValue: '1', keyType: KeyType.numeric },
      { keyValue: '0', keyType: KeyType.numeric },
      { keyValue: '.', keyType: KeyType.dot },
      { keyValue: 'C', keyType: KeyType.clear },
    ];

    const topPanelButtonDivs = this.createButtons(topPanelButtons);
    const centrePanelButtonDivs = this.createButtons(centrePanelButtons);

    return (
      <div className="CalculatorInput">
        <div className="top-panel">
          {topPanelButtonDivs}
        </div>
        <div className="centre-panel">
          {centrePanelButtonDivs}
        </div>
        <div className="bottom-panel" id="result">
          <Button
            value="="
            onClick={() => this.props.onInput({ keyValue: '=', keyType: KeyType.result })}
          />
        </div>
      </div>
    );
  }
}

export default CalculatorInput;
