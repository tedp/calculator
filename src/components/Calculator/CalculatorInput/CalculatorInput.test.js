import React from 'react';
import { shallow } from 'enzyme';
import CalculatorInput from './CalculatorInput';
import { KeyType } from '../common/enums';

describe('CalculatorInput', () => {
  const clickFn = jest.fn();

  it('should render correctly', () => {
    const component = shallow(<CalculatorInput />);

    expect(component).toMatchSnapshot();
  });

  it('button click should propogate the button value', () => {
    const component = shallow(<CalculatorInput onInput={clickFn} />);
    component
      .find('Button[value="="]')
      .simulate('click');

    expect(clickFn).toHaveBeenCalledWith({ keyValue: '=', keyType: KeyType.result });
  });
});
