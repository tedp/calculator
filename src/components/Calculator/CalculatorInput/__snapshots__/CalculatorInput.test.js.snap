// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorInput should render correctly 1`] = `
<div
  className="CalculatorInput"
>
  <div
    className="top-panel"
  >
    <Button
      key="+"
      onClick={[Function]}
      value="+"
    />
    <Button
      key="-"
      onClick={[Function]}
      value="-"
    />
    <Button
      key="×"
      onClick={[Function]}
      value="×"
    />
    <Button
      key="÷"
      onClick={[Function]}
      value="÷"
    />
  </div>
  <div
    className="centre-panel"
  >
    <Button
      key="9"
      onClick={[Function]}
      value="9"
    />
    <Button
      key="8"
      onClick={[Function]}
      value="8"
    />
    <Button
      key="7"
      onClick={[Function]}
      value="7"
    />
    <Button
      key="6"
      onClick={[Function]}
      value="6"
    />
    <Button
      key="5"
      onClick={[Function]}
      value="5"
    />
    <Button
      key="4"
      onClick={[Function]}
      value="4"
    />
    <Button
      key="3"
      onClick={[Function]}
      value="3"
    />
    <Button
      key="2"
      onClick={[Function]}
      value="2"
    />
    <Button
      key="1"
      onClick={[Function]}
      value="1"
    />
    <Button
      key="0"
      onClick={[Function]}
      value="0"
    />
    <Button
      key="."
      onClick={[Function]}
      value="."
    />
    <Button
      key="C"
      onClick={[Function]}
      value="C"
    />
  </div>
  <div
    className="bottom-panel"
    id="result"
  >
    <Button
      onClick={[Function]}
      value="="
    />
  </div>
</div>
`;
